protocol/rpcprovider/provider_state_machine.go

```diff
@@ -0,0 +1,83 @@
+package rpcprovider
+
+import (
+	"context"
+	"time"
+
+	"github.com/lavanet/lava/v2/protocol/chainlib"
+	"github.com/lavanet/lava/v2/protocol/chainlib/chainproxy/rpcclient"
+	"github.com/lavanet/lava/v2/protocol/common"
+	"github.com/lavanet/lava/v2/protocol/lavaprotocol"
+	"github.com/lavanet/lava/v2/utils"
+	pairingtypes "github.com/lavanet/lava/v2/x/pairing/types"
+)
+
+type RelaySender interface {
+	SendNodeMsg(ctx context.Context, ch chan interface{}, chainMessage chainlib.ChainMessageForSend, extensions []string) (relayReply *chainlib.RelayReplyWrapper, subscriptionID string, relayReplyServer *rpcclient.ClientSubscription, proxyUrl common.NodeUrl, chainId string, err error)
+}
+
+type ProviderStateMachine struct {
+	relayRetriesManager *lavaprotocol.RelayRetriesManager
+	chainId             string
+}
+
+func NewProviderStateMachine(chainId string) *ProviderStateMachine {
+	return &ProviderStateMachine{
+		relayRetriesManager: lavaprotocol.NewRelayRetriesManager(),
+		chainId:             chainId,
+	}
+}
+
+func (psm *ProviderStateMachine) SendNodeMessage(ctx context.Context, relaySender RelaySender, chainMsg chainlib.ChainMessage, request *pairingtypes.RelayRequest) (*chainlib.RelayReplyWrapper, error) {
+	hash, err := chainMsg.GetRawRequestHash()
+	requestHashString := ""
+	if err != nil {
+		utils.LavaFormatWarning("Failed converting message to hash", err, utils.LogAttr("url", request.RelayData.ApiUrl), utils.LogAttr("data", string(request.RelayData.Data)))
+	} else {
+		requestHashString = string(hash)
+	}
+
+	var replyWrapper *chainlib.RelayReplyWrapper
+	var isNodeError bool
+	for retryAttempt := 0; retryAttempt <= numberOfRetriesAllowedOnNodeErrors; retryAttempt++ {
+		sendTime := time.Now()
+		replyWrapper, _, _, _, _, err = relaySender.SendNodeMsg(ctx, nil, chainMsg, request.RelayData.Extensions)
+		if err != nil {
+			return nil, utils.LavaFormatError("Sending chainMsg failed", err, utils.LogAttr("attempt", retryAttempt), utils.LogAttr("GUID", ctx), utils.LogAttr("specID", psm.chainId))
+		}
+
+		if replyWrapper == nil || replyWrapper.RelayReply == nil {
+			return nil, utils.LavaFormatError("Relay Wrapper returned nil without an error", nil, utils.LogAttr("attempt", retryAttempt), utils.LogAttr("GUID", ctx), utils.LogAttr("specID", psm.chainId))
+		}
+
+		if debugLatency {
+			utils.LavaFormatDebug("node reply received", utils.LogAttr("attempt", retryAttempt), utils.LogAttr("timeTaken", time.Since(sendTime)), utils.LogAttr("GUID", ctx), utils.LogAttr("specID", psm.chainId))
+		}
+
+		// Failed fetching hash return the reply.
+		if requestHashString == "" {
+			break // We can't perform the retries as we failed fetching the request hash.
```

Comment:
this is a warning because it deteriorates performance of the provider and shouldnt happen